ihmc-java-toolkit/src/main/java/us/ihmc/tools/time/FrequencyCalculator.java

```diff
@@ -1,59 +1,96 @@
 package us.ihmc.tools.time;
 
-import us.ihmc.commons.time.Stopwatch;
+import us.ihmc.commons.thread.ThreadTools;
+import us.ihmc.log.LogTools;
 import us.ihmc.tools.UnitConversions;
 
-import java.util.ArrayDeque;
+import java.util.Deque;
+import java.util.LinkedList;
+import java.util.UUID;
 
+/**
+ * A rolling average frequency calculator with an optional logging thread to print the frequency once per second.
+ * The frequency is calculated on each query to the frequency rather than each new event.
+ * Call {@link #ping()} on each new event.
+ * Call {@link #getFrequency()} to get the current frequency.
+ */
 public class FrequencyCalculator
 {
-   private volatile double frequency;
-   private final ArrayDeque<Double> deltas = new ArrayDeque<>();
-   private final Stopwatch stopwatch = new Stopwatch().start();
-   private final int history;
-
-   /**
-    * History of 3.
-    */
-   public FrequencyCalculator()
+   private static final double NANOS_IN_A_SECOND = 1_000_000_000.0;
+
+   private double frequency;
+   private final Deque<Long> pingTimes = new LinkedList<>();
+
+   private volatile boolean loggingThreadRunning;
+
+   public FrequencyCalculator(boolean enableLoggingThread)
    {
-      this(3);
+      if (enableLoggingThread)
+      {
+         String threadID = UUID.randomUUID().toString().substring(0, 5);
+
+         Thread loggingThread = new Thread(() ->
+         {
+            loggingThreadRunning = true;
+
+            while (loggingThreadRunning)
+            {
+               LogTools.info("FrequencyCalculator[" + threadID + "] average rate: " + getFrequency());
+
+               ThreadTools.sleep(1000);
+            }
+         }, getClass().getSimpleName() + "-" + threadID);
+
+         loggingThread.start();
+      }
    }
 
-   /**
-    * History if one yields the latest calculation only.
-    * @param history 1 or more.
-    */
-   public FrequencyCalculator(int history)
+   public FrequencyCalculator()
    {
-      this.history = history;
+      this(false);
    }
 
-   public void ping()
+   private double calculateFrequency(boolean decay)
    {
-      double elapsed = stopwatch.lap();
-      deltas.addLast(elapsed);
+      Long first = pingTimes.peekFirst();
+      Long last = decay ? Long.valueOf(System.nanoTime()) : pingTimes.peekLast();
+      int pings = pingTimes.size();
 
-      while (deltas.size() > history)
+      if (first != null && last != null && pings > 1)
       {
-         deltas.removeFirst();
+         long elapsedNanos = last - first;
+         long elapsedNanosAverage = elapsedNanos / (pings - 1);
+         double elapsedSecondsAverage = elapsedNanosAverage / NANOS_IN_A_SECOND;
+         return UnitConversions.secondsToHertz(elapsedSecondsAverage);
```

Comment:
Coud do 
```
         long elapsedNanos = last - first;
         double elapsedSeconds = Conversions.nanosecondsToSeconds(elapsedNanos);
         double elapsedSecondsAverage = elapsedSeconds / (pings - 1);
         return UnitConversions.secondsToHertz(elapsedSecondsAverage);
```
Just to avoid the integer division of `elapsedNanos / (pings - 1)`. I'd only make the change if you want to be super pedantic.

Reply:
Good idea, I've applied that change